File: leetcode/course-schedule/course-schedule-dfs-iter.rb
# @param {Integer} num_courses
# @param {Integer[][]} prerequisites
# @return {Boolean}
def can_finish(num_courses, prerequisites)
  !Graph.new(num_courses, prerequisites).cyclic?
end

class Graph
  def initialize(vertex_count, edges)
    @adj = Array.new(vertex_count) { [] }
    edges.each { |src, dest| @adj[src] << dest }
  end

  def cyclic?
    vis = [:white] * @adj.size

    cyclic_from = ->(start) do
      return false if vis[start] == :black
      raise 'corrupted visitation state' unless vis[start] == :white
      vis[start] = :gray
      stack = [Frame.new(start)]

      until stack.empty?
        frame = stack.last
        row = @adj[frame.src]

        if frame.index == row.size
          vis[frame.src] = :black
          stack.pop
          next
        end

        dest = row[frame.index]
        frame.index += 1

        case vis[dest]
        when :white
          vis[dest] = :gray
          stack.push(Frame.new(dest))
        when :gray
          return true
        when :black
          next
        else
          raise 'unrecognized visitation state'
        end
      end
    end

    (0...@adj.size).any?(&cyclic_from)
  end
end

class Frame
  attr_reader :src
  attr_accessor :index

  def initialize(src)
    @src = src
    @index = 0
  end
end

File: leetcode/course-schedule/course-schedule-dfs-iter.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        var graph = new Graph(numCourses);
        for (var edge : prerequisites) graph.addEdge(edge[1], edge[0]);
        return !graph.hasCycle();
    }
}

class Graph {
    Graph(int vertexCount) {
        _adj = new ArrayList<>(vertexCount);
        while (_adj.size() < vertexCount) _adj.add(new ArrayList<>());
    }

    void addEdge(int src, int dest) {
        _adj.get(src).add(dest);
    }

    boolean hasCycle() {
        var vis = freshVisitationList();

        IntPredicate hasCycleFrom = start -> {
            if (vis[start] == Color.BLACK) return false;
            if (vis[start] != Color.WHITE)
                throw new AssertionError("corrupted visitation state");

            Deque<Frame> stack = new ArrayDeque<>();
            vis[start] = Color.GRAY;
            stack.push(new Frame(start, _adj));

            while (!stack.isEmpty()) {
                var frame = stack.peek();

                if (!frame.iter.hasNext()) {
                    vis[frame.src] = Color.BLACK;
                    stack.pop();
                    continue;
                }

                int dest = frame.iter.next();

                switch (vis[dest]) {
                case WHITE:
                    vis[dest] = Color.GRAY;
                    stack.push(new Frame(dest, _adj));
                    break;
                case GRAY:
                    return true;
                case BLACK:
                    break;
                default:
                    throw new AssertionError("unrecognized visitation state");
                }
            }

            return false;
        };

        return vertices().anyMatch(hasCycleFrom);
    }

    private Color[] freshVisitationList() {
        return vertices().mapToObj(_vertex -> Color.WHITE)
                         .toArray(Color[]::new);
    }

    private IntStream vertices() {
        return IntStream.range(0, _adj.size());
    }

    private final List<List<Integer>> _adj;
}

final class Frame {
    Frame(int src, List<List<Integer>> adj) {
        this.src = src;
        this.iter = adj.get(src).iterator();
    }

    final int src;

    final Iterator<Integer> iter;
}

enum Color {
    WHITE,
    GRAY,
    BLACK,
}
